etc/init.d/zfs-load-key.in

```diff
@@ -0,0 +1,131 @@
+#!@DEFAULT_INIT_SHELL@
+#
+# zfs-load-key  This script will load/unload the zfs filesystems keys.
+#
+# chkconfig:    2345 06 99
+# description:  This script will load or unload the zfs filesystems keys during
+#               system boot/shutdown. Only filesystems with key path set
+#               in keylocation property. See the zfs(8) man page for details.
+# probe: true
+#
+### BEGIN INIT INFO
+# Provides:          zfs-load-key
+# Required-Start:    $local_fs zfs-import
+# Required-Stop:     $local_fs zfs-import
+# Default-Start:     2 3 4 5
+# Default-Stop:      0 1 6
+# X-Start-Before:    zfs-mount
+# X-Stop-After:      zfs-zed
+# Short-Description: Load ZFS keys for filesystems and volumes
+# Description: Run the `zfs load-key` or `zfs unload-key` commands.
+### END INIT INFO
+#
+# Released under the 2-clause BSD license.
+#
+# This script is based on debian/zfsutils.zfs.init from the
+# Debian GNU/kFreeBSD zfsutils 8.1-3 package, written by Aurelien Jarno.
+
+# Source the common init script
+. @sysconfdir@/zfs/zfs-functions
+
+# ----------------------------------------------------
+
+do_depend()
+{
+	# bootmisc will log to /var which may be a different zfs than root.
+	before bootmisc logger zfs-mount
+
+	after zfs-import sysfs
+	keyword -lxc -openvz -prefix -vserver
+}
+
+# Load keys for all datasets/filesystems
+do_load_keys()
+{
+	zfs_log_begin_msg "Load ZFS filesystem(s) keys"
+
+	"$ZFS" list -Ho name,encryptionroot,keystatus,keylocation |
+		while read -r name encryptionroot keystatus keylocation; do
+			if [ "$encryptionroot" != "-" ] &&
+				[ "$name" = "$encryptionroot" ] &&
+				[ "$keystatus" = "unavailable" ] &&
+				[ "$keylocation" != "prompt" ] &&
+				[ "$keylocation" != "none" ]
+			then
+			zfs_action "Load key for $encryptionroot" \
+				"$ZFS" load-key "$encryptionroot"
+			fi
+	done
+
+	zfs_log_end_msg 0
+
+	return 0
+}
+
+# Unload keys for all datasets/filesystems
+do_unload_keys()
+{
+	zfs_log_begin_msg "Unload ZFS filesystem(s) key"
+
+	"$ZFS" list -Ho name,encryptionroot,keystatus | tac |
```

Comment:
rev. tac is very nonportable

Reply:
Yeah, I thought about that. I have no clue how to make that portable. 

From my understanding neither
`| nl | sort -nr | cut -f 2- |`
or 
` | cat -n | sort -nr | cut -f 2- |`
are very portable.

Should we use sed?
`sed '1!G;h;$!d'`

Reply:
Still not sure why you can't use rev, which is universally supported.

Reply:
rev reverses the character in every line. But I want to reverse the order of the lines (last line first).

Reply:
As I understand this should work in all POSIX environments:
`| nl | sort -nr | cut -f 2- |`

Reply:
Oh, sorry, I was so hard under the impression that `rev` reversed lines that I misread the manpage both times I looked at it!
Yeah, both `| nl | sort -nr | cut -f 2- |` and `| awk '{l[x++]=$0} END{while(x--)print l[x]}' |` will work fine here.